src/core/net/dnssd_server.cpp

```diff
@@ -1143,6 +1259,24 @@ void Server::HandleTimer(void)
         }
     }
 
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+    for (UpstreamQueryTransaction &query : mUpstreamQueryTransactions)
+    {
+        TimeMilli expire;
+
+        if (!query.IsValid())
+        {
+            continue;
+        }
+
+        expire = query.GetStartTime() + kQueryTimeout;
+        if (expire <= now)
+        {
+            query.Reset();
```

Comment:
Should we call to inform upstream to cancel the txn?

Also should we send a response to requester to say we timed out or some error?

Reply:
Makes sense, I adds a function for sending "Server Failure" here.

However, I only found SERVFAIL for "forwarding timeout" in RFC2136, which is for DNS Update, and RFC1035 does not mention this. (At least, if we want to return something, we should return SERVFAIL)

I'm thinking if this is really required, since the client will always maintain its own timer for this, and the only case we cancel a unfinished transaction here would be timeout.

Reply:
Either way is ok, we can decide not send a response and let the client time out on its own.

The good thing I see for not sending response is that then we don't need `ConstructAndSendResponseCode()` (can use the exitsing code in `ProcessQuery()` to prepare and send response).